include/swift/Sema/Constraint.h

```diff
@@ -226,6 +226,13 @@ enum class ConstraintKind : char {
   BindTupleOfFunctionParams,
   /// The first type is a type pack, and the second type is its reduced shape.
   ShapeOf,
+  /// Represents explicit generic arguments provided for a reference to
+  /// a declaration.
+  ///
+  /// The first type is the type variable describing the bound type of
+  /// an overload. The second type is a PackType containing the explicit
+  /// generic arguments.
+  ExplicitGenericArguments,
```

Comment:
A think the design with 3 types might be better - `$T_unspecialized <specialize-with> (type1, type2, ...) -> $T_specialized` because specialization would be attempted first and the rest of the constraint system would operate on the specialized type.